### PR TITLE
fix(Windows:Link): Initial fix that passes the link phase

### PR DESCRIPTION
**Pull request by RicardasN**
## Description

- Updated rename function as on Windows it required moving files, simple renaming was not enough.

## Related issues

- https://github.com/flexn-io/renative/issues/1596

## Npm releases

n/a

## Screenshots

image

### CODE REVIEW

**Review by GabrieleKaceviciute — CHANGES_REQUESTED** [review]
I've tested on Windows 10 Pro, Node v18.19.0 ->
```
C:\Users\dev\Desktop\hello-renative>rnv link

┌─────────────────────────────────────────────────────────────────┐
│ ██████╗ ███████╗███╗   ██╗ █████╗ ████████╗██╗██╗   ██╗███████╗ │
│ ██╔══██╗██╔════╝████╗  ██║██╔══██╗╚══██╔══╝██║██║   ██║██╔════╝ │
│ ██████╔╝█████╗  ██╔██╗ ██║███████║   ██║   ██║██║   ██║█████╗   │
│ ██╔══██╗██╔══╝  ██║╚██╗██║██╔══██║   ██║   ██║╚██╗ ██╔╝██╔══╝   │
│ ██║  ██║███████╗██║ ╚████║██║  ██║   ██║   ██║ ╚████╔╝ ███████╗ │
│ ╚═╝  ╚═╝╚══════╝╚═╝  ╚═══╝╚═╝  ╚═╝   ╚═╝   ╚═╝  ╚═══╝  ╚══════╝ │
│ RNV v:1.0.0-rc.19 | renative.org | 21/06/2024, 11:21:47         │
│ $ rnv link                                                      │
└─────────────────────────────────────────────────────────────────┘
info: Current engine: @rnv/engine-core (C:\Users\dev\Desktop\renative\packages\engine-core\lib)
task: ○ link ↺1
info: Found following source packages:

C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\adapter (unlinked)
C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\cli (unlinked)
C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\config-templates (unlinked)
C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\core (unlinked)
C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\engine-core (unlinked)
C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\sdk-telemetry (unlinked)
C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\sdk-utils (unlinked)
C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\template-starter (unlinked)

? Found following packages to link? @rnv/adapter, @rnv/cli, @rnv/config-templates, @rnv/core, @rnv/engine-core, @rnv/sdk
-telemetry, @rnv/sdk-utils, @rnv/template-starter
info: Linking packages...
error: ⨯  Error: EPERM: operation not permitted, symlink 'C:\Users\dev\Desktop\renative\packages\adapter' -> 'C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\adapter'
    at Object.symlinkSync (node:fs:1814:3)
    at fsSymlinkSync (C:\Users\dev\Desktop\renative\packages\core\src\system\fs.ts:59:8)
    at _linkPackage (C:\Users\dev\Desktop\renative\packages\engine-core\src\tasks\linking\taskLink.ts:36:22)
    at C:\Users\dev\Desktop\renative\packages\engine-core\src\tasks\linking\taskLink.ts:84:13
    at Array.forEach (<anonymous>)
    at C:\Users\dev\Desktop\renative\packages\engine-core\src\tasks\linking\taskLink.ts:83:34
    at step (C:\Users\dev\Desktop\renative\node_modules\tslib\tslib.js:195:27)
    at Object.next (C:\Users\dev\Desktop\renative\node_modules\tslib\tslib.js:176:57)
    at fulfilled (C:\Users\dev\Desktop\renative\node_modules\tslib\tslib.js:166:62)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)

┌─────────────────────────────────────────────────────────────────┐
│ ⨯ SUMMARY | 21/06/2024, 11:21:49 | rnv@1.0.0-rc.19              │
│ $ rnv link                                                      │
├─────────────────────────────────────────────────────────────────┤
│ Project: hello-renative@0.1.0                                   │
│ Workspace: rnv                                                  │
│ Engine: @rnv/engine-core                                        │
│ Build Scheme (-s): debug                                        │
│ Env Info: win32 | x64 | node v18.19.0                           │
│ Executed Time: 0h:0m:1s:538ms                                   │
│ error: ⨯  Error: EPERM: operation not permitted, symlink 'C:\Users\dev\Desktop\renative\packages\adapter' -> 'C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\adapter'
    at Object.symlinkSync (node:fs:1814:3)
    at fsSymlinkSync (C:\Users\dev\Desktop\renative\packages\core\src\system\fs.ts:59:8)
    at _linkPackage (C:\Users\dev\Desktop\renative\packages\engine-core\src\tasks\linking\taskLink.ts:36:22)
    at C:\Users\dev\Desktop\renative\packages\engine-core\src\tasks\linking\taskLink.ts:84:13
    at Array.forEach (<anonymous>)
    at C:\Users\dev\Desktop\renative\packages\engine-core\src\tasks\linking\taskLink.ts:83:34
    at step (C:\Users\dev\Desktop\renative\node_modules\tslib\tslib.js:195:27)
    at Object.next (C:\Users\dev\Desktop\renative\node_modules\tslib\tslib.js:176:57)
    at fulfilled (C:\Users\dev\Desktop\renative\node_modules\tslib\tslib.js:166:62)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)

└─────────────────────────────────────────────────────────────────┘
(node:424) [DEP0147] DeprecationWarning: In future versions of Node.js, fs.rmdir(path, { recursive: true }) will be removed. Use fs.rm(path, { recursive: true }) instead
(Use `node --trace-deprecation ...` to show where the warning was created)

C:\Users\dev\Desktop\hello-renative>
```

**Comment by RicardasN** [comment]
> ```
> C:\Users\dev\Desktop\hello-renative>rnv link
> 
> ┌─────────────────────────────────────────────────────────────────┐
> │ ██████╗ ███████╗███╗   ██╗ █████╗ ████████╗██╗██╗   ██╗███████╗ │
> │ ██╔══██╗██╔════╝████╗  ██║██╔══██╗╚══██╔══╝██║██║   ██║██╔════╝ │
> │ ██████╔╝█████╗  ██╔██╗ ██║███████║   ██║   ██║██║   ██║█████╗   │
> │ ██╔══██╗██╔══╝  ██║╚██╗██║██╔══██║   ██║   ██║╚██╗ ██╔╝██╔══╝   │
> │ ██║  ██║███████╗██║ ╚████║██║  ██║   ██║   ██║ ╚████╔╝ ███████╗ │
> │ ╚═╝  ╚═╝╚══════╝╚═╝  ╚═══╝╚═╝  ╚═╝   ╚═╝   ╚═╝  ╚═══╝  ╚══════╝ │
> │ RNV v:1.0.0-rc.19 | renative.org | 21/06/2024, 11:21:47         │
> │ $ rnv link                                                      │
> └─────────────────────────────────────────────────────────────────┘
> info: Current engine: @rnv/engine-core (C:\Users\dev\Desktop\renative\packages\engine-core\lib)
> task: ○ link ↺1
> info: Found following source packages:
> 
> C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\adapter (unlinked)
> C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\cli (unlinked)
> C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\config-templates (unlinked)
> C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\core (unlinked)
> C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\engine-core (unlinked)
> C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\sdk-telemetry (unlinked)
> C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\sdk-utils (unlinked)
> C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\template-starter (unlinked)
> 
> ? Found following packages to link? @rnv/adapter, @rnv/cli, @rnv/config-templates, @rnv/core, @rnv/engine-core, @rnv/sdk
> -telemetry, @rnv/sdk-utils, @rnv/template-starter
> info: Linking packages...
> error: ⨯  Error: EPERM: operation not permitted, symlink 'C:\Users\dev\Desktop\renative\packages\adapter' -> 'C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\adapter'
>     at Object.symlinkSync (node:fs:1814:3)
>     at fsSymlinkSync (C:\Users\dev\Desktop\renative\packages\core\src\system\fs.ts:59:8)
>     at _linkPackage (C:\Users\dev\Desktop\renative\packages\engine-core\src\tasks\linking\taskLink.ts:36:22)
>     at C:\Users\dev\Desktop\renative\packages\engine-core\src\tasks\linking\taskLink.ts:84:13
>     at Array.forEach (<anonymous>)
>     at C:\Users\dev\Desktop\renative\packages\engine-core\src\tasks\linking\taskLink.ts:83:34
>     at step (C:\Users\dev\Desktop\renative\node_modules\tslib\tslib.js:195:27)
>     at Object.next (C:\Users\dev\Desktop\renative\node_modules\tslib\tslib.js:176:57)
>     at fulfilled (C:\Users\dev\Desktop\renative\node_modules\tslib\tslib.js:166:62)
>     at processTicksAndRejections (node:internal/process/task_queues:95:5)
> 
> ┌─────────────────────────────────────────────────────────────────┐
> │ ⨯ SUMMARY | 21/06/2024, 11:21:49 | rnv@1.0.0-rc.19              │
> │ $ rnv link                                                      │
> ├─────────────────────────────────────────────────────────────────┤
> │ Project: hello-renative@0.1.0                                   │
> │ Workspace: rnv                                                  │
> │ Engine: @rnv/engine-core                                        │
> │ Build Scheme (-s): debug                                        │
> │ Env Info: win32 | x64 | node v18.19.0                           │
> │ Executed Time: 0h:0m:1s:538ms                                   │
> │ error: ⨯  Error: EPERM: operation not permitted, symlink 'C:\Users\dev\Desktop\renative\packages\adapter' -> 'C:\Users\dev\Desktop\hello-renative\node_modules\@rnv\adapter'
>     at Object.symlinkSync (node:fs:1814:3)
>     at fsSymlinkSync (C:\Users\dev\Desktop\renative\packages\core\src\system\fs.ts:59:8)
>     at _linkPackage (C:\Users\dev\Desktop\renative\packages\engine-core\src\tasks\linking\taskLink.ts:36:22)
>     at C:\Users\dev\Desktop\renative\packages\engine-core\src\tasks\linking\taskLink.ts:84:13
>     at Array.forEach (<anonymous>)
>     at C:\Users\dev\Desktop\renative\packages\engine-core\src\tasks\linking\taskLink.ts:83:34
>     at step (C:\Users\dev\Desktop\renative\node_modules\tslib\tslib.js:195:27)
>     at Object.next (C:\Users\dev\Desktop\renative\node_modules\tslib\tslib.js:176:57)
>     at fulfilled (C:\Users\dev\Desktop\renative\node_modules\tslib\tslib.js:166:62)
>     at processTicksAndRejections (node:internal/process/task_queues:95:5)
> 
> └─────────────────────────────────────────────────────────────────┘
> (node:424) [DEP0147] DeprecationWarning: In future versions of Node.js, fs.rmdir(path, { recursive: true }) will be removed. Use fs.rm(path, { recursive: true }) instead
> (Use `node --trace-deprecation ...` to show where the warning was created)
> 
> C:\Users\dev\Desktop\hello-renative>
> ```

Cleaned everything and cannot reproduce the error, am I doing anything wrong?
https://drive.google.com/file/d/1Yo4znrb1niCKpOEGnqOuulVNktjvJvdR/view?usp=drive_link

**Comment by GabrieleKaceviciute** [comment]
@RicardasN steps seems the same ->

```
C:\Users\dev\Desktop>rnv new

┌─────────────────────────────────────────────────────────────────┐
│ ██████╗ ███████╗███╗   ██╗ █████╗ ████████╗██╗██╗   ██╗███████╗ │
│ ██╔══██╗██╔════╝████╗  ██║██╔══██╗╚══██╔══╝██║██║   ██║██╔════╝ │
│ ██████╔╝█████╗  ██╔██╗ ██║███████║   ██║   ██║██║   ██║█████╗   │
│ ██╔══██╗██╔══╝  ██║╚██╗██║██╔══██║   ██║   ██║╚██╗ ██╔╝██╔══╝   │
│ ██║  ██║███████╗██║ ╚████║██║  ██║   ██║   ██║ ╚████╔╝ ███████╗ │
│ ╚═╝  ╚═╝╚══════╝╚═╝  ╚═══╝╚═╝  ╚═╝   ╚═╝   ╚═╝  ╚═══╝  ╚══════╝ │
│ RNV v:1.0.0-rc.19 | renative.org | 21/06/2024, 11:20:57         │
│ $ rnv new                                                       │
└─────────────────────────────────────────────────────────────────┘
info: Current engine: @rnv/engine-core (C:\Users\dev\Desktop\renative\packages\engine-core\lib)
task: ○ new ↺1
? What's your project Name? (folder will be created) hello-renative
? What workspace to use? rnv
? What template to use? @rnv/template-starter (local) - \Users\dev\Desktop\renative\packages\template-starter
info: Found local template: @rnv/template-starter@1.0.0-rc.19
√ Executing: yarn
√ Executing: yarn add @rnv/config-templates@1.0.0-rc.19 --dev
? How to create config renative.json? Extend template (cleaner, overridable)
? What's your project Title? My Renative App
? What's your App ID? com.mycompany.hellorenative
? What's your project version? 0.1.0
? What platforms would you like to use? web, ios, android
? Do you want to set-up git in your new project? Yes
info: Your project does not have a git repo. Creating one...DONE
√ Executing: git init
info: Missing directory C:\Users\dev\Desktop\hello-renative\buildHooks.js. COPYING from TEMPLATE...DONE
info: Missing directory C:\Users\dev\Desktop\hello-renative\src\app.js. COPYING from TEMPLATE...DONE
info: Missing directory C:\Users\dev\Desktop\hello-renative\src\entry.js. COPYING from TEMPLATE...DONE
info: Missing file C:\Users\dev\Desktop\hello-renative\src\config.tsx.js. COPYING from TEMPLATE...DONE
info: Missing file C:\Users\dev\Desktop\hello-renative\index.js.js. COPYING from TEMPLATE...DONE
info: Missing file C:\Users\dev\Desktop\hello-renative\tsconfig.json.js. COPYING from TEMPLATE...DONE
info: Missing file C:\Users\dev\Desktop\hello-renative\babel.config.js.js. COPYING from TEMPLATE...DONE
info: Missing directory C:\Users\dev\Desktop\hello-renative\typings.js. COPYING from TEMPLATE...DONE
info: Missing file C:\Users\dev\Desktop\hello-renative\.gitignore.js. COPYING from TEMPLATE...DONE
info: Missing directory C:\Users\dev\Desktop\hello-renative\appConfigs\app.js. COPYING from TEMPLATE...DONE
info: Missing directory C:\Users\dev\Desktop\hello-renative\appConfigs\base\assets.js. COPYING from TEMPLATE...DONE
info: Missing directory C:\Users\dev\Desktop\hello-renative\appConfigs\base\fonts.js. COPYING from TEMPLATE...DONE
info: Missing file C:\Users\dev\Desktop\hello-renative\Gemfile.js. COPYING from TEMPLATE...DONE
info: Missing file C:\Users\dev\Desktop\hello-renative\metro.config.js.js. COPYING from TEMPLATE...DONE
info: Missing file C:\Users\dev\Desktop\hello-renative\react-native.config.js.js. COPYING from TEMPLATE...DONE
info: Missing file C:\Users\dev\Desktop\hello-renative\next.config.js.js. COPYING from TEMPLATE...DONE
info: Missing file C:\Users\dev\Desktop\hello-renative\next-env.d.ts.js. COPYING from TEMPLATE...DONE
info: Missing directory C:\Users\dev\Desktop\hello-renative\src\pages.js. COPYING from TEMPLATE...DONE
info: .rnv/schemas folder missing. CREATING...DONE
info: Updating appConfig app with title: My Renative App and id: com.mycompany.hellorenative
? Project is ready. Do you want to install all dependencies now? Yes
√ Executing: yarn
task: ✔ new
┌─────────────────────────────────────────────────────────────────┐
│ ✔ SUMMARY | 21/06/2024, 11:21:36 | rnv@1.0.0-rc.19              │
│ $ rnv new                                                       │
├─────────────────────────────────────────────────────────────────┤
│ Project: hello-renative@0.1.0                                   │
│ Workspace: rnv                                                  │
│ Engine: @rnv/engine-core                                        │
│ Template: @rnv/template-starter                                 │
│ Build Scheme (-s): debug                                        │
│ Env Info: win32 | x64 | node v18.19.0                           │
│ Executed Time: 0h:0m:38s:550ms                                  │
│
  Generated Project Summary:
  -------------------------
  Project Name (--projectName): hello-renative
  Package name: hello-renative
  Project Version (--appVersion): 0.1.0
  Workspace (--workspace): rnv
  App Title (--title): My Renative App
  App ID (--id): com.mycompany.hellorenative
  Project Template (--template): \Users\dev\Desktop\renative\packages\template-starter
  Git Enabled (--gitEnabled): true
  Enabled Platforms: web, ios, android
  -------------------------
  ✔ Your project is ready! Run it with:

  cd hello-renative
  npx rnv run
└─────────────────────────────────────────────────────────────────┘

C:\Users\dev\Desktop>cd hello-renative
```

```
C:\Users\dev\Desktop\renative>git status
On branch fix/windows-link
Your branch is up to date with 'origin/fix/windows-link'.
```

**Comment by pauliusguzas** [comment]
On my personal win pc works, minion same result as @GabrieleKaceviciute

**Review by GabrieleKaceviciute — APPROVED** [review]

File: packages/core/src/system/fs.ts [hunk 1]
```diff
@@ -37,7 +37,17 @@ export const fsReadFileSync = (dest: fs.PathLike | undefined) => fs.readFileSync
...
 
-export const fsRenameSync = (arg1: fs.PathLike | undefined, arg2: fs.PathLike) => fs.renameSync(arg1!, arg2);
+export const fsRenameSync = (arg1: fs.PathLike | undefined, arg2: fs.PathLike) => {
+    // One of the paths does not exist
+    if (!arg1 || !arg2) return;
+
+    if (fs.lstatSync(arg1).isDirectory()) {
+        fs.cpSync(arg1 as string, arg2 as string, { recursive: true });
+        fs.rmdirSync(arg1, { recursive: true });
+    } else if (fs.lstatSync(arg1).isFile()) {
```
Comment (Marius456) [inline packages/core/src/system/fs.ts:47]: Wouldn't it be better to have default function result for example throw an error when ifs can't catch unexpected conditions.

**Review by Marius456 — CHANGES_REQUESTED** [review]

File: packages/core/src/system/fs.ts (hunk 1 above)
Comment (RicardasN) [reply packages/core/src/system/fs.ts:47]: Do you mean adding a trycatch block or having some sort of separate returns for inside if conditions and then an error return?

**Review by Marius456 — APPROVED** [review]

**Pull request merged**